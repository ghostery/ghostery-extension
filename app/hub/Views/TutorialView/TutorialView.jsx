/**
 * Tutorial View Component
 *
 * Ghostery Browser Extension
 * https://www.ghostery.com/
 *
 * Copyright 2018 Ghostery, Inc. All rights reserved.
 *
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0
 */

import React from 'react';
import PropTypes from 'prop-types';
import { Route } from 'react-router-dom';

// Components
import TutorialNavigation from '../TutorialViews/TutorialNavigation';

/**
 * A Functional React component for rendering the Tutorial View
 * @return {JSX} JSX for rendering the Setup View of the Hub app
 * @memberof HubComponents
 */
const TutorialView = props => (
	<div className="full-height flex-container flex-dir-column">
		<div className="flex-child-grow">
			{props.steps.map(step => (
				<Route
					key={`route-${step.index}`}
					path={step.path}
					render={() => (
						<div>
							<step.bodyComponent />
						</div>
					)}
				/>
			))}
		</div>

		<TutorialNavigation steps={props.steps} />
	</div>
);

// PropTypes ensure we pass required props of the correct type
TutorialView.propTypes = {
	steps: PropTypes.arrayOf(PropTypes.shape({
		index: PropTypes.number.isRequired,
		path: PropTypes.string.isRequired,
		bodyComponent: PropTypes.func.isRequired,
	})).isRequired,
};

export default TutorialView;
