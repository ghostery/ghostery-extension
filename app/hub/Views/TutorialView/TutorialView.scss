/**
 * Tutorial View Sass
 *
 * Ghostery Browser Extension
 * https://www.ghostery.com/
 *
 * Copyright 2018 Ghostery, Inc. All rights reserved.
 *
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0
 *
 * ToDo: Update to BEM [Block Element Modifier](https://csswizardry.com/2013/01/mindbemding-getting-your-head-round-bem-syntax/)
 */

// purple rectangle outline
.rectangle-1 {
	width: 120px;
	height: 43px;
	border-radius: 7.2px;
	box-shadow: 0 2px 4px 0 rgba(0, 0, 0, 0.5);
	border: solid 2.7px #930194;
	transform: scale(0);
	-webkit-animation: zoom .5s;
	animation: zoom .5s;
	-webkit-animation-fill-mode: forwards;
	animation-fill-mode: forwards;
	-webkit-animation-delay: 1s;
	animation-delay: 1s;
}


@-webkit-keyframes zoom {
	100% {
		-webkit-transform: scale(1,1);
	}
}

@keyframes zoom {
	100% {
		transform: scale(1,1);
	}
}
