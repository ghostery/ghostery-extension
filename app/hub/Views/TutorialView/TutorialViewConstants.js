/**
 * Tutorial View Constants
 *
 * Ghostery Browser Extension
 * https://www.ghostery.com/
 *
 * Copyright 2018 Ghostery, Inc. All rights reserved.
 *
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0
 */

export const INIT_TUTORIAL_PROPS = 'INIT_TUTORIAL_PROPS';

export const SET_TUTORIAL_NAVIGATION = 'SET_TUTORIAL_NAVIGATION';
