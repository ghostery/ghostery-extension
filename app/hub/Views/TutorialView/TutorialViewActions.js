/**
 * Tutorial View Action creators
 *
 * Ghostery Browser Extension
 * https://www.ghostery.com/
 *
 * Copyright 2018 Ghostery, Inc. All rights reserved.
 *
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0
 */

import { INIT_TUTORIAL_PROPS, SET_TUTORIAL_NAVIGATION } from './TutorialViewConstants';

export function initTutorialProps(data) {
	return function (dispatch) {
		return new Promise((resolve) => {
			dispatch({
				type: INIT_TUTORIAL_PROPS,
				data,
			});
			resolve();
		});
	};
}

export function setTutorialNavigation(data) {
	return {
		type: SET_TUTORIAL_NAVIGATION,
		data,
	};
}
