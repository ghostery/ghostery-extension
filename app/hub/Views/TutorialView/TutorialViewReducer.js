/**
 * Reducer used throughout the Tutorial View's flow
 *
 * Ghostery Browser Extension
 * https://www.ghostery.com/
 *
 * Copyright 2018 Ghostery, Inc. All rights reserved.
 *
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0
 */
import { INIT_TUTORIAL_PROPS, SET_TUTORIAL_NAVIGATION } from './TutorialViewConstants';

const initialState = {};

function TutorialViewReducer(state = initialState, action) {
	switch (action.type) {
		// Tutorial View
		case INIT_TUTORIAL_PROPS: {
			const {
				activeIndex,
				hrefPrev,
				hrefNext,
				hrefDone,
				textPrev,
				textNext,
				textDone,
			} = action.data.navigation;
			return Object.assign({}, state, {
				tutorial: {
					navigation: {
						activeIndex,
						hrefPrev,
						hrefNext,
						hrefDone,
						textPrev,
						textNext,
						textDone,
					}
				},
			});
		}
		case SET_TUTORIAL_NAVIGATION: {
			const {
				activeIndex,
				hrefPrev,
				hrefNext,
				hrefDone,
				textPrev,
				textNext,
				textDone,
			} = action.data;
			return Object.assign({}, state, {
				tutorial: Object.assign({}, state.tutorial, {
					navigation: {
						activeIndex,
						hrefPrev,
						hrefNext,
						hrefDone,
						textPrev,
						textNext,
						textDone,
					},
				}),
			});
		}

		default: return state;
	}
}

export default TutorialViewReducer;
