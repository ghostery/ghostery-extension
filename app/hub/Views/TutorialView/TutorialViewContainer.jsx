/**
 * Tutorial View Container
 *
 * Ghostery Browser Extension
 * https://www.ghostery.com/
 *
 * Copyright 2018 Ghostery, Inc. All rights reserved.
 *
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0
 */

import React from 'react';
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';

import TutorialView from './TutorialView';
import * as actions from './TutorialViewActions';

// Component Views
import TutorialVideoView from '../TutorialViews/TutorialVideoView';
import TutorialTrackerListView from '../TutorialViews/TutorialTrackerListView';
import TutorialSimpleDetailedView from '../TutorialViews/TutorialSimpleDetailedView';
import TutorialBlockingView from '../TutorialViews/TutorialBlockingView';
import TutorialTrustRestrictView from '../TutorialViews/TutorialTrustRestrictView';
import TutorialAntiSuiteView from '../TutorialViews/TutorialAntiSuiteView';

/**
 * @class Implement the Tutorial View Container for the Ghostery Hub
 * @extends Component
 * @memberof HubContainers
 */
class TutorialViewContainer extends React.Component {
	/**
	 * Lifecycle Event
	 */
	componentWillMount() {
		const title = '';
		window.document.title = title;
	}

	/**
	 * React's required render function. Returns JSX
	 * @return {JSX} JSX for rendering the Tutorial View of the Hub app
	 */
	render() {
		const activeIndex = +this.props.location.pathname.split('/').pop();
		const steps = [
			{
				index: 1,
				path: '/tutorial/1',
				bodyComponent: TutorialVideoView,
			},
			{
				index: 2,
				path: '/tutorial/2',
				bodyComponent: TutorialTrackerListView,
			},
			{
				index: 3,
				path: '/tutorial/3',
				bodyComponent: TutorialSimpleDetailedView,
			},
			{
				index: 4,
				path: '/tutorial/4',
				bodyComponent: TutorialBlockingView,
			},
			{
				index: 5,
				path: '/tutorial/5',
				bodyComponent: TutorialTrustRestrictView,
			},
			{
				index: 6,
				path: '/tutorial/6',
				bodyComponent: TutorialAntiSuiteView,
			},
		];

		return <TutorialView activeIndex={activeIndex} steps={steps} />;
	}
}

// Default props used throughout the Tutorial flow
TutorialViewContainer.defaultProps = {};

/**
 * Map redux store state properties to the component's own properties.
 * @param  {Object} state    entire Redux store's state
 * @return {function}        this function returns a plain object, which will be merged into the component's props
 * @memberof HubContainers
 */
const mapStateToProps = () => Object.assign({});

/**
 * Bind the component's action creators using Redux's bindActionCreators.
 * @param  {function} dispatch redux store method which dispatches actions
 * @return {function}          to be used as an argument in redux connect call
 * @memberof SetupContainers
 */
const mapDispatchToProps = dispatch => ({
	actions: bindActionCreators(Object.assign(actions), dispatch),
});

export default connect(mapStateToProps, mapDispatchToProps)(TutorialViewContainer);
