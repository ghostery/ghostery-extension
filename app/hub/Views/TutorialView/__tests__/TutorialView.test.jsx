/**
 * Tutorial View Test Component
 *
 * Ghostery Browser Extension
 * https://www.ghostery.com/
 *
 * Copyright 2018 Ghostery, Inc. All rights reserved.
 *
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0
 */

import React from 'react';
import renderer from 'react-test-renderer';
import { shallow } from 'enzyme';
import { MemoryRouter } from 'react-router';
import TutorialView from '../TutorialView';

// Mock Necessary Imports
jest.mock('../../TutorialViews/TutorialNavigation', () => props => <div>Mock Tutorial Navigation</div>);

// Mock Test Component
const TestComponent = props => <div>test component</div>;

describe('app/hub/Views/TutorialView component', () => {
	describe('Snapshot tests with react-test-renderer', () => {
		test('tutorial view is rendered correctly', () => {
			const initialState = {
				steps: [
					{
						index: 0,
						path: '',
						bodyComponent: TestComponent,
					},
					{
						index: 1,
						path: '',
						bodyComponent: TestComponent,
					},
				],
				sendMountActions: true,
			};

			const component = renderer.create(
				<MemoryRouter>
					<TutorialView {...initialState} />
				</MemoryRouter>
			).toJSON();
			expect(component).toMatchSnapshot();
		});
	});

	describe('Shallow snapshot tests rendered with Enzyme', () => {
		test('the happy path of the component', () => {
			const initialState = {
				steps: [],
				sendMountActions: true,
			};

			const component = shallow(<TutorialView {...initialState} />);
			expect(component.find('.TutorialView--paddingTopLarge').length).toBe(1);

			expect(component.find('[totalSteps=0]').length).toBe(1);
			expect(component.find('[totalSteps=2]').length).toBe(0);
			component.setProps({ steps: [
				{
					index: 0,
					path: '',
					bodyComponent: TestComponent,
				},
				{
					index: 1,
					path: '',
					bodyComponent: TestComponent,
				},
			] });
			expect(component.find('[totalSteps=0]').length).toBe(0);
			expect(component.find('[totalSteps=2]').length).toBe(1);
		});
	});
});
