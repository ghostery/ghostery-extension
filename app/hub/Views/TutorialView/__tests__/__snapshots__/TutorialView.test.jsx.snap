// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`app/hub/Views/TutorialView component Snapshot tests with react-test-renderer tutorial view is rendered correctly 1`] = `
<div
  className="full-height flex-container flex-dir-column"
>
  <div
    className="TutorialView--paddingTopLarge flex-child-grow flex-container align-middle"
  >
    <div>
      test component
    </div>
    <div>
      test component
    </div>
  </div>
  <div>
    Mock Tutorial Navigation
  </div>
</div>
`;
